ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -706,104 +717,151 @@ chainSyncClient mkPipelineDecision0 tracer cfg
+                      theirFrag               = theirFrag'
+                    , theirHeaderStateHistory = theirHeaderStateHistory'
+                    , ourFrag                 = ourFrag
+                    , mostRecentIntersection  = mostRecentIntersection'
+                    }
+            atomically $ writeTVar varCandidate theirFrag'
+            atomically $ traceWith headerMetricsTracer (slotNo, now)
+
+            continueWithState kis'' $ nextStep mkPipelineDecision n theirTip
+
+    -- Used in 'rollForward': determines whether the header is from the future,
+    -- and handle that fact if so. Also return the ledger state used for the
+    -- determination.
+    --
+    -- Relies on 'readLedgerState'.
+    checkArrivalTime :: KnownIntersectionState blk
+                     -> arrival
+                     -> EarlyExit.WithEarlyExit m (Intersects blk (LedgerState blk))
+    checkArrivalTime kis arrival = do
+        Intersects kis' (lst, judgment) <- readLedgerState kis $ \lst ->
+          case runExcept $ judgeHeaderArrival (configLedger cfg) lst arrival of
+            Left PastHorizon{} -> Nothing
+            Right judgment     -> Just (lst, judgment)
+
+        -- For example, throw an exception if the header is from the far
+        -- future.
+        EarlyExit.lift $ handleHeaderArrival judgment >>= \case
+          Just exn -> disconnect (InFutureHeaderExceedsClockSkew exn)
+          Nothing  -> return $ Intersects kis' lst
+
+    -- Used in 'rollForward': block until the the ledger state at the
+    -- intersection with the local selection returns 'Just'.
+    --
+    -- Exits early if the intersection no longer exists.
+    readLedgerState :: KnownIntersectionState blk
+                    -> (LedgerState blk -> Maybe a)
+                    -> EarlyExit.WithEarlyExit m (Intersects blk a)
+    readLedgerState kis prj = join $ EarlyExit.lift $ atomically $ do
+        -- We must first find the most recent intersection with the current
+        -- chain. Note that this is cheap when the chain and candidate haven't
+        -- changed.
         mKis' <- intersectsWithCurrentChain kis
         case mKis' of
-          Nothing -> return NoLongerIntersects
+          Nothing -> return EarlyExit.exitEarly
           Just kis'@KnownIntersectionState { mostRecentIntersection } -> do
-            -- We're calling 'ledgerViewForecastAt' in the same STM transaction
-            -- as 'intersectsWithCurrentChain'. This guarantees the former's
-            -- precondition: the intersection is within the last @k@ blocks of
-            -- the current chain.
-            forecast <-
+            lst <-
               maybe
                 (error $
                    "intersection not within last k blocks: " <> show mostRecentIntersection)
-                (ledgerViewForecastAt (configLedger cfg) . ledgerState)
+                ledgerState
                 <$> getPastLedger mostRecentIntersection
```

Comment:
I was first confused by the `join` as this function doesn't actually perform any monadic effects apart from the `atomically`, everything is in `STM`, we just use `join` for `WithEarlyExit`.

It could be rewritten like this
```haskell
    readLedgerState kis prj = atomically $ do
        -- We must first find the most recent intersection with the current
        -- chain. Note that this is cheap when the chain and candidate haven't
        -- changed.
        kis'@KnownIntersectionState { mostRecentIntersection } <-
          EarlyExit.earlyExit $ intersectsWithCurrentChain kis
        lst <- EarlyExit.lift $
          maybe
            (error $
               "intersection not within last k blocks: " <> show mostRecentIntersection)
            ledgerState
            <$> getPastLedger mostRecentIntersection
        case prj lst of
          Nothing         -> retry
          Just ledgerView -> pure $ Intersects kis' ledgerView
```
by making use of the fact that
```haskell
type STM (WithEarlyExit m) = WithEarlyExit (STM m)
```
Wrinkle is that `EarlyExit.earlyExit` is not exported:
https://github.com/input-output-hk/ouroboros-consensus/blob/6614998b855d19e947bd9d7a166458a2baea66d4/ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Util/EarlyExit.hs#L60-L62
Seems like a rather natural function to me (could also be renamed to `exitOnNothing`), and of course doesn't actually add any power.

Reply:
I think I prefer the way it is, with the `join`. I pushed up a commit that reveals an intermediate type signature---is that commit enough to make the type shenanigans easier to follow?

Reply:
I still feel that
```haskell
    -- Used in 'rollForward': block until the the ledger state at the
    -- intersection with the local selection returns 'Just'.
    --
    -- Exits early if the intersection no longer exists.
    readLedgerState :: KnownIntersectionState blk
                    -> (LedgerState blk -> Maybe a)
                    -> EarlyExit.WithEarlyExit m (Intersects blk a)
    readLedgerState kis prj = atomically $ do
        -- We must first find the most recent intersection with the current
        -- chain. Note that this is cheap when the chain and candidate haven't
        -- changed.
        EarlyExit.lift (intersectsWithCurrentChain kis) >>= \case
          Nothing   -> EarlyExit.exitEarly
          Just kis' -> do
            let KnownIntersectionState { mostRecentIntersection } = kis'
            lst <- EarlyExit.lift $
              maybe
                (error $
                   "intersection not within last k blocks: " <> show mostRecentIntersection)
                ledgerState
                <$> getPastLedger mostRecentIntersection
            case prj lst of
              Nothing         -> retry
              Just ledgerView -> pure $ Intersects kis' ledgerView
```
is strictly more direct, but this is not a major point (especially with the intermediate type signature, it should be hard to be confused also with the current approach, despite being a bit verbose).